usaspending_api/references/management/commands/load_naics.py

```diff
@@ -1,48 +1,74 @@
-    DEFAULT_FILEPATH = os.path.join(DEFAULT_DIRECTORY,  'naics_codes.csv')
+    path = 'usaspending_api/data/naics_archive'
+    path = os.path.normpath(path)
+    default_path = os.path.join(settings.BASE_DIR, path)
 
     def add_arguments(self, parser):
-            parser.add_argument(
-                '-f', '--file',
-                default=self.DEFAULT_FILEPATH,
-                help='path to CSV file to load',
-            )
+        parser.add_argument('-p', '--path', help='the path to the Excel spreadsheets to load', default=self.default_path)
+        parser.add_argument('-a', '--append', help='Append to existing guide', action='store_true', default=True)
 
     def handle(self, *args, **options):
+        load_naics(path=options['path'], append=options['append'])
+
+
+@transaction.atomic
+def load_naics(path, append):
+
+    if append:
+        logging.info('Appending definitions to existing guide')
+    else:
+        logging.info('Deleting existing definitions from guide')
+        NAICS.objects.all().delete()
+
+    # year regex object precompile
+    p_year = re.compile("(20[0-9]{2})")
+
+    dir_files = glob.glob(path + "/*.xlsx")
+
+    for fname in os.listdir(path):
+
+        if fname.find("naics") > 0:
+            dir_files.append(os.path.join(path, fname))
+
+    for path in dir_files:
+        wb = load_workbook(filename=path)
+        ws = wb.active
+
+        naics_year = p_year.search(ws["A1"].value).group()
+
+        for current_row, row in enumerate(ws.rows):
+            if current_row == 0:
+                continue
+
+            if not row[0].value:
+                break  # Reads file only until a blank line
+
+            naics_code = row[0].value
+            naics_desc = row[1].value
+
+            obj, created = NAICS.objects.get_or_create(pk=naics_code)
 
-        filepath = options['file']
-        fullpath = os.path.join(django.conf.settings.BASE_DIR, filepath)
-
-        load_naics(fullpath)
-        self.logger.log(20, "Loading NAICS complete.")
-
-
-def load_naics(fullpath):
-    """
-    Create CFDA Program records from a CSV of historical data.
-    """
-    try:
-        with open(fullpath, errors='backslashreplace', encoding='utf-8-sig') as csvfile:
-
-            reader = csv.DictReader(csvfile, delimiter=',', quotechar='"', skipinitialspace='true')
-            for row in reader:
-                naics, created = NAICS.objects.get_or_create(
-                                code=row['2017 NAICS Code'])
-                naics.description = row['2017 NAICS Title'].strip()
-                naics.save()
-    except IOError:
-        logger = logging.getLogger('console')
-        logger.log("Could not open file to load from")
+            if not created:
+                if int(naics_year) > int(obj.year):
+                    NAICS.objects.filter(pk=naics_code).update(description=naics_desc, year=naics_year)
+            else:
```

Comment:
Am I understanding this properly?
If `created is True`, update the description and year

Is that necessary?

Reply:
`created` would be `True` a new entry is created as obj and the description and year is added and database updated. otherwise, if the NAICS code already exists, it would only update if the current entry is more up-to-date

Reply:
It doesn't make sense to run two write operations.

`NAICS.objects.get_or_create(pk=naics_code, defaults={'description':naics_desc, 'year':naics_year})` should write the entire record the first time. Then you don't need to update the description and year.

Reply:
Perfect, thanks, didn't know the defaults parameter existed.